### PR TITLE
[NVPTX] support dynamic allocas with PTX alloca instruction

### PR DESCRIPTION
**Pull request by AlexMaclean**
Add support for dynamically sized alloca instructions with the PTX alloca instruction introduced in PTX 7.3 ([9.7.15.3. Stack Manipulation Instructions: alloca](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#stack-manipulation-instructions-alloca))

### CODE REVIEW

**Comment by elhewaty** [comment]
I was about to work on this issue https://github.com/llvm/llvm-project/issues/62812, I think your patch will fix it.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td [hunk 1]
```diff
@@ -3805,6 +3805,28 @@ def CALL_PROTOTYPE :
...
+  SDNode<"NVPTXISD::DYNAMIC_STACKALLOC", SDTDynAllocaOp,
+         [SDNPHasChain, SDNPSideEffect]>;
+
+def DYNAMIC_STACKALLOC32 :
+  NVPTXInst<(outs Int32Regs:$ptr),
+            (ins Int32Regs:$size, i32imm:$align),
+            "alloca.u32 \t$ptr, $size, $align;\n\t"
+            "cvta.local.u32 \t$ptr, $ptr;",
+            [(set (i32 Int32Regs:$ptr), (dyn_alloca Int32Regs:$size, (i32 timm:$align)))]>,
+            Requires<[hasPTX<73>]>;
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3821]: also `hasSM<52>`

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td [hunk 2]
```diff
@@ -3805,6 +3805,28 @@ def CALL_PROTOTYPE :
...
+
+def DYNAMIC_STACKALLOC64 :
+  NVPTXInst<(outs Int64Regs:$ptr),
+            (ins Int64Regs:$size, i32imm:$align),
+            "alloca.u64 \t$ptr, $size, $align;\n\t"
+            "cvta.local.u64 \t$ptr, $ptr;",
+            [(set Int64Regs:$ptr, (dyn_alloca Int64Regs:$size, (i32 timm:$align)))]>,
+            Requires<[hasPTX<73>]>;
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3829]: ditto.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 3]
```diff
@@ -2211,14 +2210,24 @@ SDValue NVPTXTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
...
+  else
+    Size = DAG.getZExtOrTrunc(Size, DL, MVT::i32);
+
+  SDValue AllocOps[] = {Chain, Size,
+                        DAG.getTargetConstant(Align, DL, MVT::i32)};
+  SDValue Alloca = DAG.getNode(NVPTXISD::DYNAMIC_STACKALLOC, DL,
+                               nvTM->is64Bit() ? MVT::i64 : MVT::i32, AllocOps);
+
+  SDValue MergeOps[] = {Alloca, Chain};
+  return DAG.getMergeValues(MergeOps, DL);
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2230]: We still need to diagnose it in some cases, as `alloca` instruction is only available on sm_52+ and PTX7.3+

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 1 above)
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Fixed

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 2 above)
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Fixed

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 3 above)
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2245]: Makes sense, I've added the diagnose code back

**Comment by AlexMaclean** [comment]
> I was about to work on this issue #62812, I think your patch will fix it.

I just gave compiling the provided IR a try and I can confirm that with this change it is now successfully compiled (provided a supported arch and PTX version is specified).

**Comment by elhewaty** [comment]
That's great. BTW do you have a TODO list I am interested in nvptx but I don't know where to start?

**Comment by AlexMaclean** [comment]
> That's great. BTW do you have a TODO list I am interested in nvptx but I don't know where to start?

@elhewaty, Sorry, I don't personally have anything like this. I'll ask my team, but I don't think NVIDIA has any official list or documentation of this form either. If you have any ideas I'd be happy to give my 2 cents on whether or not I think they make sense to work on.

**Comment by AlexMaclean** [comment]
@Artem-B, I've address your comments, does this look ready to land now?

**Review by Artem-B — APPROVED** [review]

**Pull request merged**